### PR TITLE
Heartbeat event format

### PR DESCRIPTION
**Pull request by urso**
PR addresses #3406 and #3838.

Additional work on extracting meta-data from connection layers should be done in follow-up PR.

Work done:
- add look.Status
- update event format per monitor
- Rewrite heartbeat dialchain package:
  - simplify package
  - have connection layers add standardized connection data to the final event
  - add some more helpers
  - add some godocs

## Event format changes:

| Old Field | Changes |
| --- | --- |
| monitor | -> monitor.id |
| type | <ul><li>-> monitor.name</li><li>defaults to monitor.type if not configured</li><li>introduce new `type` fields with default 'monitor' until `type` is not required anymore</li></ul>|
| duration | -> monitor.duration |
|scheme | -> monitor.scheme |
|up | <ul><li> -> monitor.status </li><li>- replace boolean with keywords 'up', 'down' </li></ul>|
| host | <ul><li>-> monitor.host</li><li> -> resolve.host </li></ul> |
| ip | <ul><li>-> monitor.ip</li><li> -> resolve.ip </li></ul> |
|resolve_rtt | -> resolve.rtt |
| icmp_rtt | -> icmp.rtt |
| port | -> tcp.port
| tcp_connect_rtt | -> tcp.rtt.connect |
| validate_rtt | -> tcp.rtt.validate |
| socks5_connect_rtt | -> socks5.rtt.connect |
| tls_handshake_rtt | -> tls.rtt.handshake |
| url | -> http.url |
| response.status | ->  http.response.status |
| rtt | -> http.rtt.total |
| http_rtt | -> http.rtt.response_header |

New fields:
- `type`: always set to `"monitor"`. Can be removed when not required by libbeat anymore
- `monitor.type`: the monitor type, e.g. "tcp", "icmp", "http".
- `http.rtt.validate`: Duration between first byte of HTTP request being written and response being processed by validator. Duration based on already available network connection.
- `http.rtt.validate_body`: Duration of validator required to read and validate the response
- `http.rtt.write_request`: Duration of writing the complete HTTP request. Duration based on already available network connection.

## Sample Events:

ICMP Event:

```
{
  "@timestamp": "2017-04-24T11:36:53.639Z",
  "beat": {
    "hostname": "mbp",
    "name": "mbp",
    "version": "6.0.0-alpha1"
  },
  "icmp": {
    "requests": 1,
    "rtt": {
      "us": 87
    }
  },
  "monitor": {
    "duration": {
      "us": 2557
    },
    "host": "localhost",
    "ip": "127.0.0.1",
    "name": "myname",
    "status": "up",
    "type": "icmp"
  },
  "resolve": {
    "host": "localhost",
    "ip": "127.0.0.1",
    "rtt": {
      "us": 2316
    }
  }
}
```

ICMP event for unknown host:

```
{
  "@timestamp": "2017-04-24T11:36:53.639Z",
  "beat": {
    "hostname": "mbp",
    "name": "mbp",
    "version": "6.0.0-alpha1"
  },
  "error": {
    "message": "lookup unknown-host: no such host",
    "type": "io"
  },
  "monitor": {
    "duration": {
      "us": 7628
    },
    "host": "unknown-host",
    "name": "myname",
    "status": "down",
    "type": "icmp"
  },
  "resolve": {
    "host": "unknown-host"
  }
}
```

TCP Event:

```
{
  "@timestamp": "2017-04-24T12:29:04.649Z",
  "beat": {
    "hostname": "mbp",
    "name": "mbp",
    "version": "6.0.0-alpha1"
  },
  "monitor": {
    "duration": {
      "us": 4461
    },
    "host": "localhost",
    "ip": "127.0.0.1",
    "name": "tcp",
    "scheme": "tcp", // Note: can be 'tcp' or 'tls' for TCP monitor. HTTP monitor sets 'http' or 'https'
    "status": "up",
    "type": "tcp"
  },
  "resolve": {
    "host": "localhost",
    "ip": "127.0.0.1",
    "rtt": {
      "us": 4136
    }
  },
  "tcp": {
    "port": "6060",
    "rtt": {
      "connect": {
        "us": 183
      }
    }
  }
}
```

TCP + TLS Event:

```
{
  "@timestamp": "2017-04-24T12:29:04.649Z",
  "beat": {
    "hostname": "mbp",
    "name": "mbp",
    "version": "6.0.0-alpha1"
  },
  "monitor": {
    "duration": {
      "us": 234591
    },
    "host": "play.golang.org",
    "ip": "216.58.213.241",
    "name": "tcp",
    "scheme": "tls",
    "status": "up",
    "type": "tcp"
  },
  "resolve": {
    "host": "play.golang.org",
    "ip": "216.58.213.241",
    "rtt": {
      "us": 51663
    }
  },
  "tcp": {
    "port": "443",
    "rtt": {
      "connect": {
        "us": 46140
      }
    }
  },
  "tls": {
    "rtt": {
      "handshake": {
        "us": 136605
      }
    }
  }
}
```

HTTP Event with 404:

```
{
  "@timestamp": "2017-04-24T13:44:44.718Z",
  "beat": {
    "hostname": "mbp",
    "name": "mbp",
    "version": "6.0.0-alpha1"
  },
  "error": {
    "message": "404 Not Found",
    "type": "validate"
  },
  "http": {
    "response": {
      "status": 404
    },
    "rtt": {
      "content": {
        "us": 244
      },
      "response_header": {
        "us": 225
      },
      "total": {
        "us": 537
      },
      "write_request": {
        "us": 63
      }
    },
    "url": "http://localhost:6060"
  },
  "monitor": {
    "duration": {
      "us": 4935
    },
    "host": "localhost",
    "ip": "127.0.0.1",
    "name": "http",
    "scheme": "http",
    "status": "down",
    "type": "http"
  },
  "resolve": {
    "host": "localhost",
    "ip": "127.0.0.1",
    "rtt": {
      "us": 4271
    }
  },
  "tcp": {
    "port": "6060",
    "rtt": {
      "connect": {
        "us": 247
      }
    }
  }
}
```

HTTPS event:

```
{
  "@timestamp": "2017-04-24T13:44:44.718Z",
  "beat": {
    "hostname": "mbp",
    "name": "mbp",
    "version": "6.0.0-alpha1"
  },
  "http": {
    "response": {
      "status": 200
    },
    "rtt": {
      "content": {
        "us": 167523
      },
      "response_header": {
        "us": 167433
      },
      "total": {
        "us": 339594
      },
      "write_request": {
        "us": 102
      }
    },
    "url": "https://play.golang.org"
  },
  "monitor": {
    "duration": {
      "us": 344043
    },
    "host": "play.golang.org",
    "ip": "216.58.213.241",
    "name": "http",
    "scheme": "https",
    "status": "up",
    "type": "http"
  },
  "resolve": {
    "host": "play.golang.org",
    "ip": "216.58.213.241",
    "rtt": {
      "us": 4272
    }
  },
  "tcp": {
    "port": "443",
    "rtt": {
      "connect": {
        "us": 38350
      }
    }
  },
  "tls": {
    "rtt": {
      "handshake": {
        "us": 133618
      }
    }
  }
}
```

### CODE REVIEW

File: heartbeat/_meta/fields.yml [hunk 1]
```diff
@@ -2,131 +2,257 @@
...
+                      number does not fully represent the total time needed
+                      to read the body.
+              fields:
+                - name: us
+                  type: long
+                  description: Duration in microseconds
+
+            - name: validate_body
+              type: group
+              description: |
```
Comment (ruflin) [inline heartbeat/_meta/fields.yml]: Normally we use a `>`. I assume `|` also works?

File: heartbeat/monitors/active/http/task.go [hunk 2]
```diff
@@ -9,6 +9,7 @@ import (
 	"net/http"
 	"strconv"
 	"strings"
+
```
Comment (ruflin) [inline heartbeat/monitors/active/http/task.go]: empty line too much?

File: heartbeat/monitors/active/http/task.go [hunk 3]
```diff
@@ -41,25 +42,25 @@ func newHTTPMonitorHostJob(
 		return nil, err
 	}
 
-	hostname, port, err := splitHostnamePort(request)
+	hostname, _, err := splitHostnamePort(request)
```
Comment (ruflin) [inline heartbeat/monitors/active/http/task.go]: port is excluded because it is already in the tcp layer, right?

File: heartbeat/monitors/util.go [hunk 4]
```diff
@@ -53,108 +75,114 @@ func (s IPSettings) Network() string {
...
-
 	return func(ip *net.IPAddr) TaskRunner {
 		cont := f(ip)
 		switch len(cont) {
 		case 0:
 			return emptyTask
-		case 1:
-			return makeTask(cont[0])
+		case
+			1:
```
Comment (ruflin) [inline heartbeat/monitors/util.go]: is the  `1:` intentional on a new line?

File: heartbeat/monitors/util.go [hunk 5]
```diff
@@ -293,30 +383,46 @@ func (p *PingMode) Unpack(s string) error {
...
 }
 
-func annotated(start time.Time, typ string, fn func() (common.MapStr, []TaskRunner, error)) TaskRunner {
+func annotated(
+	settings JobSettings,
+	start time.Time,
+	fn func() (common.MapStr, []TaskRunner, error),
+) TaskRunner {
 	return MakeCont(func() (common.MapStr, []TaskRunner, error) {
+		fmt.Println("run task")
```
Comment (ruflin) [inline heartbeat/monitors/util.go]: leftover?

File: heartbeat/monitors/util.go [hunk 6]
```diff
@@ -293,30 +383,46 @@ func (p *PingMode) Unpack(s string) error {
...
+
 		event, cont, err := fn()
+		fmt.Println("task returns: ", event, cont, err)
```
Comment (ruflin) [inline heartbeat/monitors/util.go]: left over?

File: heartbeat/monitors/util.go [hunk 7]
```diff
@@ -293,30 +383,46 @@ func (p *PingMode) Unpack(s string) error {
...
+
 		if err != nil {
 			if event == nil {
 				event = common.MapStr{}
 			}
+			fmt.Println("set error: ", err)
```
Comment (ruflin) [inline heartbeat/monitors/util.go]: left over?

File: heartbeat/monitors/util.go [hunk 8]
```diff
@@ -293,30 +383,46 @@ func (p *PingMode) Unpack(s string) error {
...
 			event["error"] = look.Reason(err)
 		}
 
 		if event != nil {
-			event.Update(common.MapStr{
+			status := look.Status(err)
+			fmt.Println("set status: ", status)
```
Comment (ruflin) [inline heartbeat/monitors/util.go]: left over?

File: heartbeat/_meta/fields.yml (hunk 1 above)
Comment (urso) [inline heartbeat/_meta/fields.yml]: oh, oversight from my side. in YAML one can use `>`, `>-` or `|`.

`>`: folded-style. Replace newlines with space. Multiple blank lines are replaced with newline (or space in some implementations)
`|`: literal block. Keeps newlines and blank lines.

I wanted to enforce the newline (or better blank line?) before "Note:" in the hopes of the asciidoc putting the note in a separate new line. That's why I used `|`. 

If I remember correctly `>-` replaces a blank line (`\n\n`) by one newline (`\n`).

File: heartbeat/monitors/active/http/task.go (hunk 2 above)
Comment (urso) [inline heartbeat/monitors/active/http/task.go]: interesting. so gimport woes here? Will fix.

File: heartbeat/monitors/active/http/task.go (hunk 3 above)
Comment (urso) [inline heartbeat/monitors/active/http/task.go]: Port is missing here. This constructor is used when a proxy is configured + name lookup happens by proxy itself. I did undo the tcp dialer to report the port number due to potential use of proxies.

File: heartbeat/monitors/util.go (hunk 4 above)
Comment (urso) [inline heartbeat/monitors/util.go]: ups.

File: heartbeat/monitors/util.go (hunk 5 above)
Comment (urso) [inline heartbeat/monitors/util.go]: Thanks!

**Pull request merged**